exonum/src/runtime/rust/service.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::runtime::{error::ExecutionError, rust::TransactionContext, MethodId};
+
+use failure::Error;
+
+pub trait ServiceDispatcher {
+    fn call(
+        &self,
+        method: MethodId,
+        ctx: TransactionContext,
+        payload: &[u8],
+    ) -> Result<Result<(), ExecutionError>, Error>;
+}
+
+pub trait SystemService: ServiceDispatcher + std::fmt::Debug {
+    // constructor method
+    // other hooks such as "on node startup"
+}
+
+#[macro_export]
+macro_rules! service_interface {
```

Comment:
A pure magic. I guess it's OK for now, but later we need to document this thing very carefully.